test/unit-tests/function/algebra/simplifyConstant.test.js

```diff
@@ -22,9 +22,47 @@ describe('simplifyConstant', function () {
     testSimplifyConstant('0.5 x', '1 / 2 x')
   })
 
-  it('should coalesce constants in a multi-argument expression', function () {
-    testSimplifyConstant('3 + x + 7 + y', '10 + x + y')
-    testSimplifyConstant('3 * x * 7 * y', '21 * x * y')
+  describe('should coalesce constants in a multi-argument expression', function () {
+    it('in the default context', function () {
+      testSimplifyConstant('3 + x + 7 + y', '10 + x + y')
+      testSimplifyConstant('3 * x * 7 * y', '21 * x * y')
+
+      // including an 'unevaluable' constant
+      // todo?: output here not currently as expected; instead outputting as '9 + "foo" + 11'.
+      // testSimplifyConstant('2 + 7 + "foo" + 3 + 8', '20 + "foo"')
+    })
+
+    it('in non-commutative contexts', function () {
+      const context = {
+        add: { commutative: false },
+        multiply: { commutative: false }
+      }
+      const opts = { context }
+
+      // leading consts.
+      testSimplifyConstant('2 + 2 + a', '4 + a', undefined, opts)
+      testSimplifyConstant('2 * 2 * b', '4 * b', undefined, opts)
+      // trailing consts.
+      testSimplifyConstant('a + 3 + 2', 'a + 5', undefined, opts)
+      testSimplifyConstant('a * 3 * 2', 'a * 6', undefined, opts)
+      // 'wedged' constants
+      testSimplifyConstant('a + 3 + 2 + a + 4 + 4 + b', 'a + 5 + a + 8 + b', undefined, opts)
+      testSimplifyConstant('c * 3 * 2 * d * 4 * 1 * d', 'c * 6 * d * 4 * d', undefined, opts)
+
+      // including an 'unevaluable' constant
+      testSimplifyConstant('2 + 7 + "foo" + 3 + 8', '9 + "foo" + 11', undefined, opts)
+      // collapsing of constants with a 'falsy' value (e.g. '0')
+      testSimplifyConstant('a + 0 + 7 + b', 'a + 7 + b', undefined, opts)
+
+      // note:
+      // - tests with inverses of commutative operations. will not be as expected (i.e. 'fully
+      // simplified') if passed directly to simplifyConstant, and further they should not be needed.
+      // In the absence of prior application of standard rules which flatten substraction and division
+      // ops. (such as rule 'n / n1 -> n * n1 ^ (-1)' for div. and 'n - n1 -> n + -n1' for
+      // subtraction), constant args are not 'flattened' and consequently included in 'multi-arg
+      // expressions', therefore prohibiting evaluation of adjacent constants (in a non-commutative
+      // context) here.
+    })
```

Comment:
Rather than or in addition to all of these comments (which I will be honest I had a little trouble following), maybe you should do some concrete examples as tests, so that the reader will see what simplifyConstant does do (and we will be verifying it across future changes).

Reply:
Hi gwhitney. I do not believe that it makes sense to add tests for these cases. Instead, I have re-written the commentary elaborating on why these tests are absent (in some sense a 'limitation' on using simplifyConstant alone).
I'll push up the changes now, including the enabling of a test as per your previous comment, but you are free to add a test or two in addition regarding this if you still feel that it is warranted.

Reply:
Up to @josdejong but an essay of this length now feels out of place in the code, even in the tests; perhaps you could move this to one of the issues/discussions on Github? If Jos is ok with it in the test code, then of course that's fine.

Reply:
Thanks for thinking all of this through Samuel. I think your comments intent to be a "TODO" or "FIXME", but I find it hard to distill the exact action point to be addressed to resolve the concern. Should it be a TODO on the level of `simplify` instead of `simplifyConstant` since there are some cross-concerns? Would it be possible to write a "placeholder" test like Glen suggests? Something like:

```js
// TODO: implement support for ...
// testSimplifyConstant('...', '.', undefined, opts)
// testSimplifyConstant('...', '.', undefined, opts)
```

Reply:
TODO tests added. Let me know what you think regards where these additional accumulated cases should be placed. Both of the issues there are present within simplifyConstant and across all contexts, so perhaps a (backlogged) issue centered on this function 👍

Reply:
Ah, yes that is perfect! Thanks a lot. 

It would be good I think to open a new discussion indeed to discuss how to implement support for that. Can you take the initiative to open this discussion?

I'll mark this conversation as resolved now and will merge the PR.